source/shared/cpp/ObjectModel/TextBlock.cpp

```diff
@@ -165,3 +168,141 @@ void TextBlock::SetHorizontalAlignment(const HorizontalAlignment value)
+    std::ostringstream parsedostr;
+    time_t offset = 0; 
+    bool isDate = false, isShort = true;
+    enum MatchIndex
+    {
+        IsDate = 1,
+        YEAR,
+        Month,
+        Day,
+        Hour,
+        Min,
+        Sec,
+        TimeZone,
+        TZHr = 10,
+        TZMn,
+        Format,
+    };
+    int factor = 1, hours = 0, minutes = 0;
+    struct tm parsedTm = { 0 };
+    int *addrs[] = {&parsedTm.tm_year, &parsedTm.tm_mon, 
+        &parsedTm.tm_mday, &parsedTm.tm_hour, &parsedTm.tm_min, 
+        &parsedTm.tm_sec, &hours, &minutes}; 
+    std::vector<int> indexer = {YEAR, Month, Day, Hour, Min, Sec, TZHr, TZMn};
+
+    while(std::regex_search(text, matches, pattern))
+    {
+        parsedostr << matches.prefix().str();
+
+        if(matches[IsDate].matched)
+        {
+            // Date is matched
+            isDate = (matches[IsDate].str()[0] == 'D')? true : false;
+        }
+
+        for(unsigned int idx = 0; idx < indexer.size(); idx++) 
+        { 
+            if(matches[indexer[idx]].matched)
+            {
+                // get indexes for time attributes to index into conrresponding matches
+                // and covert it to string
+                *addrs[idx] = stoi(matches[indexer[idx]]);
+            }
+        }
+
+        // maches offset sign, 
+        // Z == UTC, 
+        // + == time added from UTC
+        // - == time subtracted from UTC
+        if(matches[TimeZone].matched)
+        {
+            
+            char zone = matches[TimeZone].str()[0];
+            if(zone == 'Z')
+            {
+                factor = 0;
+            }
+            else if(zone == '+')
+            {
+                factor = -1;
+            }
+        }
+
+        // measured from year 1900
+        parsedTm.tm_year -= 1900;
+        parsedTm.tm_mon  -= 1;
+
+        // converts to seconds
+        hours *= 3600;
+        minutes *= 60;
+
+        // time zone offset calculation
+        offset = (hours + minutes) * factor;
+
+        time_t utc;
+        // converts to ticks in UTC
+        utc = mktime(&parsedTm);
+        if(utc == -1)
+        {
+            parsedostr << matches[0];
+        }
+
+        char tzOffsetBuff[6] = { 0 };
+        // gets local time zone offset
+        strftime(tzOffsetBuff, 6, "%z", &parsedTm);
+        std::string localTimeZoneOffsetStr(tzOffsetBuff);
+        int nTzOffset = std::stoi(localTimeZoneOffsetStr);
+        offset += ((nTzOffset / 100) * 3600 + (nTzOffset % 100) * 60);
+        // add offset to utc
+        utc += offset;
+        struct tm result = {0};
+
+        // match for long/short
+        if(matches[Format].matched)
+        {
+           char format = matches[Format].str()[0];
+           isShort = (format != ')' && matches[Format].str()[1] == 'L')? false : true;
+        }
+
+        // converts to local time from utc
+        if(!localtime_s(&result, &utc))
+        {
+            // localtime_s double counts daylight saving time
```

Comment:
>localtime_s double counts daylight saving time [](start = 15, length = 46)

Is this a known bug in localtime_s?

Reply:
there is in localtime
it uses global variable.

---
In reply to: [149179117](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r149179117) [](ancestors = 149179117)

Reply:
Does the same thing apply for localtime_r?

---
In reply to: [149179338](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r149179338) [](ancestors = 149179338,149179117)

Reply:
No, it doe not

---
In reply to: [149517346](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r149517346) [](ancestors = 149517346,149179338,149179117)

Reply:
localtime_r doesn't have bug

---
In reply to: [150949291](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r150949291) [](ancestors = 150949291,149517346,149179338,149179117)

Reply:
Wouldn't this mean we'd be one hour off in the platforms that use localtime_r?

---
In reply to: [150949653](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r150949653) [](ancestors = 150949653,150949291,149517346,149179338,149179117)

Reply:
no, that's incorrect; i will fix the comments to make it clear. thanks

---
In reply to: [150950310](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r150950310) [](ancestors = 150950310,150949653,150949291,149517346,149179338,149179117)